### DISCUSSION TITLE
Memory leak on iOS Safari

### DISCUSSION THREAD

**Discussion by mango-tree** [vikejs/vike#1788] [Discussion]
Hi, I'm a developer working in [Cookieplace](https://github.com/cookieplace). We're using Vike in our service, [Crepe](https://crepe.cm).
We received a report that our service was stuttering and running hot on iOS Safari, and after investigating, we realized that the stuttering and hotness was caused by memory piling up and the GC running continuously. 
When I checked why the memory is piling up, I saw in Vike that the memory is piling up in the history related logic.

So, I would like to ask two questions.

1. looking at this image, I see that the window objects are piling up, and I would like to know if this is related to the memory leak in Vike.
<img width="686" alt="스크린샷 2024-08-05 오후 6 52 23" src="https://github.com/user-attachments/assets/c3aad935-54c3-493a-8639-d3f4387a898e">

2. if it is a memory leak, I was wondering if you could fix it.

**Comment by brillout** [comment]
First of all, thanks for sponsoring :green_heart: it makes a big difference!

> the memory leak in Vike

What makes you think the memory leak is because of Vike?

**Reply by mango-tree** [reply]
When I comment `pushHistoryState()`, `replaceHistoryState()`, `monkeyPatchHistoryPushState()` functions' logic in `vike/dist/esm/client/client-routing-runtime/history.js`, there is no memory leak

**Reply by mango-tree** [reply]
I made pageContext WeakMap to check how many pageContext has been collected, then I found that Chromium keeps 3~8 pageContexts, but Safari keeps more than 30 pageContexts. Is this can make performance issues?

**Reply by brillout** [reply]
Let me have a look whether I can see a potential memory leak in these functions.

**Reply by brillout** [reply]
> window objects are piling up

Can you elaborate on that?

> When I comment `pushHistoryState()`, `replaceHistoryState()`, `monkeyPatchHistoryPushState()` functions' logic in `vike/dist/esm/client/client-routing-runtime/history.js`, there is no memory leak

- If you only comment out `pushHistoryState()` and `replaceHistoryState()` does it fix the memory leak?
- If you only comment out `monkeyPatchHistoryPushState()` does it fix the memory leak?

> I found that Chromium keeps 3~8 pageContexts, but Safari keeps more than 30 pageContexts

What do you mean with "keeps pageContexts"? Do you mean with the states passed to `window.history.pushState(state, '', url)`?

**Reply by mango-tree** [reply]
Sorry for late reply.
> If you only comment out pushHistoryState() and replaceHistoryState() does it fix the memory leak?

yes, I found that the memory is not increase too much when I capture memory with `console.takeHeapSnapshot()`.

> If you only comment out monkeyPatchHistoryPushState() does it fix the memory leak?

No, it happens again.

> What do you mean with "keeps pageContexts"? Do you mean with the states passed to window.history.pushState(state, '', url)?

The variables which define in `renderPageNominal()`

```
  async function renderPageNominal() {
    const onError = async (err: unknown) => {
      await renderErrorPage({ err })
    }

    const pageContext = await getPageContextBegin()
    ...
```

**Reply by brillout** [reply]
> > If you only comment out pushHistoryState() and replaceHistoryState() does it fix the memory leak?
> 
> yes, I found that the memory is not increase too much when I capture memory with `console.takeHeapSnapshot()`.
> 
> > If you only comment out monkeyPatchHistoryPushState() does it fix the memory leak?
> 
> No, it happens again.

Maybe it's an issue with iOS Safari? Your anlysis, and the fact that this only happens with iOS Safari, is an indication that the issue resides on iOS Safari's side. (Or maybe some other tool in your stack that monkey patches `window.history.pushState()` or `window.history.replaceState()`.)

Does iOS Safari has known memory issues with the History API?

> I found that Chromium keeps 3~8 pageContexts, but Safari keeps more than 30 pageContexts [...] The variables which define in `renderPageNominal()`

That shouldn't happen. Can you reproduce this with a new fresh and minmal Vike app?

Also, can you reproduce the memory issue with a minimal Vike app?

**Reply by mango-tree** [reply]
> Maybe it's an issue with iOS Safari? Your anlysis, and the fact that this only happens with iOS Safari, is an indication that the issue resides on iOS Safari's side. (Or maybe some other tool in your stack that monkey patches window.history.pushState() or window.history.replaceState().)

I have an issue with iOS Safari & OS X Safari (in M2 MacBook).

> Does iOS Safari has known memory issues with the History API?

I don't know whether Safari has known memory issues. I searched Webkit Bugzilla, but I didn't find it.

> That shouldn't happen. Can you reproduce this with a new fresh and minmal Vike app?
Also, can you reproduce the memory issue with a minimal Vike app?

I can't reproduce from vike example. I'll make minimal app, but it would take time. I use `react-relay` on the website, and this might be the cause.

**Reply by brillout** [reply]
> I can't reproduce from vike example. I'll make minimal app, but it would take time. I use `react-relay` on the website, and this might be the cause.

Maybe it's unrelated to Vike then.

**Comment by mango-tree** [comment]
I solved it by update Vike packages & refactor route structure. Thanks for investigating with us!

**Reply by brillout** [reply]
I'm glad you resolved the issue! Thank you for sponsoring :green_heart: it makes a real difference.

**Reply by brillout** [reply]
Hello @mango-tree, I saw that @cookieplace stopped sponsoring. Did you stop using Vike or just stop sponsoring? Best regards, Rom
